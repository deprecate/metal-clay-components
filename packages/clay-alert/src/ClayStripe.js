import './ClayAlertBase';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayStripe.soy.js';

/**
 * Metal Clay Stripe component.
 */
class ClayStripe extends Component {
	/**
	 * @inheritDoc
	 */
	attached() {
		this.addListener('hide', this.defaultHideStripe_, true);
	}

	/**
	 * Hides the alert and destroy it if proceed.
	 * @private
	 */
	defaultHideStripe_() {
		if (this.destroyOnHide) {
			this.dispose();
		}
	}

	/**
	 * Continues the propagation of the hide event
	 */
	handleHide_() {
		return !this.emit('hide');
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayStripe.STATE = {
	/**
	 * Flag to indicate if alert should be automatically closed.
	 * @instance
	 * @memberof ClayStripe
	 * @type {?bool}
	 * @default false
	 */
	autoClose: Config.bool().value(false),

	/**
	 * Flag to indicate if the alert should be destroyen when close.
	 * @instance
	 * @memberof ClayStripe
	 * @type {?bool}
	 * @default false
	 */
	destroyOnHide: Config.bool().value(false),

	/**
	 * The message of alert
	 * @instance
	 * @memberof ClayStripe
	 * @type {!html|string}
	 */
	message: Config.any().required(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayIcon
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * The style of alert
	 * @instance
	 * @memberof ClayStripe
	 * @type {?string}
	 * @default info
	 */
	style: Config.oneOf(['danger', 'info', 'success', 'warning']).value('info'),

	/**
	 * The title of alert
	 * @instance
	 * @memberof ClayStripe
	 * @type {!string}
	 * @default undefined
	 */
	title: Config.string().required(),
};

defineWebComponent('clay-stripe', ClayStripe);

Soy.register(ClayStripe, templates);

export {ClayStripe};
export default ClayStripe;
