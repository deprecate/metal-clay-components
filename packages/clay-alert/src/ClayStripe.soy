{namespace ClayStripe}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param message: html|string}
	{@param spritemap: string}
	{@param title: string}
	{@param? autoClose: bool}
	{@param? handleHide_: any}
	{@param? style: string}

  {call ClayAlertBase.render}
    {param autoClose: $autoClose /}
    {param closeable: true /}
    {param events: ['hide': $handleHide_ ] /}
    {param message: $message /}
    {param ref: 'alertBase' /}
    {param spritemap: $spritemap /}
    {param style: $style /}
    {param title: $title /}
    {param type: 'stripe' /}
  {/call}
{/template}