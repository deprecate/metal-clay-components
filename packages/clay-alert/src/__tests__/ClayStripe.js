import ClayStripe from '../ClayStripe';

let stripe;

const spritemap = '../node_modules/lexicon-ux/build/images/icons/icons.svg';

describe('ClayStripe', function() {
	afterEach(() => {
		if (stripe) {
			stripe.dispose();
		}
	});

	it('should render default markup', function() {
		stripe = new ClayStripe({
			message: 'message',
			spritemap: spritemap,
			title: 'info',
		});

		expect(stripe).toMatchSnapshot();
	});

	it('should render a "danger" style stripe', function() {
		stripe = new ClayStripe({
			message: 'message',
			spritemap: spritemap,
			style: 'danger',
			title: 'Danger!',
		});

		expect(stripe).toMatchSnapshot();
	});

	it('should render an "info" style stripe', function() {
		stripe = new ClayStripe({
			message: 'message',
			spritemap: spritemap,
			style: 'info',
			title: 'Info',
		});

		expect(stripe).toMatchSnapshot();
	});

	it('should render a "success" style stripe', function() {
		stripe = new ClayStripe({
			message: 'message',
			spritemap: spritemap,
			style: 'success',
			title: 'Success!',
		});

		expect(stripe).toMatchSnapshot();
	});

	it('should render a "warning" style stripe', function() {
		stripe = new ClayStripe({
			message: 'message',
			spritemap: spritemap,
			style: 'warning',
			title: 'Warning!',
		});

		expect(stripe).toMatchSnapshot();
	});

	it('should close an stripe', function() {
		stripe = new ClayStripe({
			message: 'message',
			spritemap: spritemap,
			title: 'Success!',
		});

		stripe.element.querySelector('button.close').click();

		expect(stripe.visible_).toBeFalsy();
	});

	// eslint-disable-next-line
	it('should close automatically an stripe after 5 seconds when message has no links', function() {
		jest.useFakeTimers();

		stripe = new ClayStripe({
			autoClose: true,
			message: 'message',
			spritemap: spritemap,
			title: 'Success!',
		});

		expect(setTimeout.mock.calls.length).toBe(1);
		expect(setTimeout.mock.calls[0][1]).toBe(5000);

		jest.runAllTimers();
		expect(stripe.visible_).toBeFalsy();
	});

	// eslint-disable-next-line
	it('should close automatically an stripe after 5 seconds when message has links', function() {
		jest.useFakeTimers();

		stripe = new ClayStripe({
			autoClose: true,
			message: 'message<a href="#">Link</a>',
			spritemap: spritemap,
			title: 'Success!',
		});

		expect(setTimeout.mock.calls.length).toBe(1);
		expect(setTimeout.mock.calls[0][1]).toBe(10000);

		jest.runAllTimers();
		expect(stripe.visible_).toBeFalsy();
	});

	it('should close and destroy an stripe', function() {
		stripe = new ClayStripe({
			destroyOnHide: true,
			message: 'message',
			spritemap: spritemap,
			title: 'Success!',
		});

		stripe.element.querySelector('button.close').click();

		expect(stripe.visible_).toBeFalsy();
		expect(stripe.disposed_).toBeTruthy();
	});

	it('should fail when no message is passed', function() {
		expect(() => {
			stripe = new ClayStripe({
				spritemap: spritemap,
				title: 'title',
			});
		}).toThrow();
	});

	it('should fail when no spritemap is passed', function() {
		expect(() => {
			stripe = new ClayStripe({
				message: 'message',
				title: 'title',
			});
		}).toThrow();
	});

	it('should fail when no title is passed', function() {
		expect(() => {
			stripe = new ClayStripe({
				message: 'message',
				spritemap: spritemap,
			});
		}).toThrow();
	});
});
