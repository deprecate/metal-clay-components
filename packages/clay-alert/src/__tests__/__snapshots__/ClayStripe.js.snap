// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayStripe should render a "danger" style stripe 1`] = `
<div class="alert alert-danger fade alert-dismissible alert-fluid show" role="alert">
  <div class="container">
    <button class="btn close btn-unstyled" aria-label="Close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
      </svg>
    </button>
    <span class="alert-indicator">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-exclamation-full">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#exclamation-full"></use>
      </svg>
    </span>
    <strong class="lead">Danger!</strong>message</div>
</div>
`;

exports[`ClayStripe should render a "success" style stripe 1`] = `
<div class="alert alert-success fade alert-dismissible alert-fluid show" role="alert">
  <div class="container">
    <button class="btn close btn-unstyled" aria-label="Close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
      </svg>
    </button>
    <span class="alert-indicator">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-check-circle-full">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#check-circle-full"></use>
      </svg>
    </span>
    <strong class="lead">Success!</strong>message</div>
</div>
`;

exports[`ClayStripe should render a "warning" style stripe 1`] = `
<div class="alert alert-warning fade alert-dismissible alert-fluid show" role="alert">
  <div class="container">
    <button class="btn close btn-unstyled" aria-label="Close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
      </svg>
    </button>
    <span class="alert-indicator">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-warning-full">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#warning-full"></use>
      </svg>
    </span>
    <strong class="lead">Warning!</strong>message</div>
</div>
`;

exports[`ClayStripe should render an "info" style stripe 1`] = `
<div class="alert alert-info fade alert-dismissible alert-fluid show" role="alert">
  <div class="container">
    <button class="btn close btn-unstyled" aria-label="Close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
      </svg>
    </button>
    <span class="alert-indicator">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
      </svg>
    </span>
    <strong class="lead">Info</strong>message</div>
</div>
`;

exports[`ClayStripe should render an stripe with classes 1`] = `
<div class="alert alert-info fade alert-dismissible my-custom-class alert-fluid show" role="alert">
  <div class="container">
    <button class="btn close btn-unstyled" aria-label="Close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
      </svg>
    </button>
    <span class="alert-indicator">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
      </svg>
    </span>
    <strong class="lead">info</strong>message</div>
</div>
`;

exports[`ClayStripe should render an stripe with id 1`] = `
<div class="alert alert-info fade alert-dismissible alert-fluid show" id="myId" role="alert">
  <div class="container">
    <button class="btn close btn-unstyled" aria-label="Close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
      </svg>
    </button>
    <span class="alert-indicator">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
      </svg>
    </span>
    <strong class="lead">info</strong>message</div>
</div>
`;

exports[`ClayStripe should render default markup 1`] = `
<div class="alert alert-info fade alert-dismissible alert-fluid show" role="alert">
  <div class="container">
    <button class="btn close btn-unstyled" aria-label="Close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
      </svg>
    </button>
    <span class="alert-indicator">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
      </svg>
    </span>
    <strong class="lead">info</strong>message</div>
</div>
`;
